org.eclipse.jface.text/src/org/eclipse/jface/text/source/inlined/InlinedAnnotationSupport.java

```diff
@@ -485,16 +490,15 @@ public void updateAnnotations(Set<AbstractInlinedAnnotation> annotations) {
 	 */
 	@SuppressWarnings("unchecked")
 	public <T extends AbstractInlinedAnnotation> T findExistingAnnotation(Position pos) {
```

Comment:
And that makes me wonder whether this method is correct at all, even in the current/previous state: I don't think there is any guarantee that there is only 1 annotation for 1 position, is there?

Reply:
From a user's point of view, in my opinion, it doesn't make sense to have multiple code annotations on the same range position or if the annotations are overlapping like so Pos1 = [0,1,2], Pos2 = [2,3,4].
I've created an example, I've modified the "InlinedAnnotationDemo" class example to show the overlapping of annotations:
![img](https://user-images.githubusercontent.com/120033779/206503403-feeef4fe-b574-435c-81c9-6aade7e65e7f.png)
Also to prove my point:
- check CodeMiningManager.renderCodeMinings() line 256, where you can clearly see that you can't have annotations on the same position, if you correctly use this mechanism and let the CodeMiningManager manage the annotations.
- in the InlinedAnnotationDemo.java you can also see that they tried to mimic the same behavior as above, where it verified if there is already an annotation for a position.
- the InlinedAnnotationSupport.updateAnnotations() method is called only from CodeMiningManager.renderCodeMinings() which guarantees that there is going to be just 1 annotation and 1 position.

I would say that if it's still not the case that it should be a 1:1 pair it should be made clear in which direction this code mining feature is going.
-Allow multiple annotations on the same position: then you would have to adapt the code and make it understandable that is possible.
-Not allow multiple annotations on the same position: then make it more restrictive if it's needed and document it.

Reply:
OK, so I believe the current code has a bug and a design flaw as there is no string reason to make an a position can only have 1 annotation. In practice, annotations can come from different providers and the case of identical positions is probable and not really hard to deal with and the API specification never excluded that.
But I get that your patch is an improvement over current state and that it doesn't exclude to fix the API later. So you can ignore that comment and we'll review the change as it without considering other pre-existing issue.